File: src/services/issueService.ts
import prisma from '../db';
import logger from '../utils/logger';
import { Issue, IssueScope, Link, Severity, IssueType, IssueState } from '@prisma/client';

export interface IssueCreateInput {
  title: string;
  description: string;
  severity: Severity;
  issueType: IssueType;
  state?: IssueState;
  namespace: string;
  scope: {
    resourceType: string;
    resourceName: string;
  };
  links?: {
    title: string;
    url: string;
  }[];
}

export interface IssueUpdateInput {
  title?: string;
  description?: string;
  severity?: Severity;
  issueType?: IssueType;
  state?: IssueState;
  resolvedAt?: Date | null;
  links?: {
    title: string;
    url: string;
  }[];
}

export interface IssueQueryFilters {
  namespace?: string;
  severity?: Severity;
  issueType?: IssueType;
  state?: IssueState;
  resourceType?: string;
  resourceName?: string;
  search?: string;
  limit?: number;
  offset?: number;
}

export interface IssueWithRelations extends Issue {
  scope: IssueScope;
  links: Link[];
  relatedIssues?: Issue[];
}

class IssueService {
  /**
   * Find issues with optional filters
   */
  async findIssues(filters: IssueQueryFilters): Promise<{
    data: IssueWithRelations[];
    total: number;
    limit: number;
    offset: number;
  }> {
    try {
      const {
        namespace,
        severity,
        issueType,
        state,
        resourceType,
        resourceName,
        search,
        limit = 50,
        offset = 0,
      } = filters;

      // Build filter object
      const where: any = {
        namespace: namespace as string,
        ...(issueType ? { issueType } : {}),
        ...(severity ? { severity } : {}),
        ...(state ? { state } : {}),
        ...(resourceType ? { scope: { resourceType }} : {}),
        ...(resourceName ? { scope: { resourceName }} : {}),
        ...(search ? {
          OR: [
            { title: { contains: search } },
            { description: { contains: search }}
          ]
        } : {})
      };

      const issues = await prisma.issue.findMany({
        where,
        include: {
          scope: true,
          links: true,
          relatedFrom: {
            include: {
              target: {
                include: {
                  scope: true
                }
              }
            }
          },
          relatedTo: {
            include: {
              source: {
                include: {
                  scope: true
                }
              }
            }
          }
        },
        orderBy: {
          detectedAt: 'desc'
        },
        skip: Number(offset),
        take: Number(limit)
      });

      // Get total count for pagination
      const total = await prisma.issue.count({ where });

      return {
        data: issues as unknown as IssueWithRelations[],
        total,
        limit: Number(limit),
        offset: Number(offset)
      };
    } catch (error) {
      logger.error(`Error in findIssues: ${error}`);
      throw error;
    }
  }

  /**
   * Find a single issue by ID
   */
  async findIssueById(id: string): Promise<IssueWithRelations | null> {
    try {
      const issue = await prisma.issue.findUnique({
        where: { id },
        include: {
          scope: true,
          links: true,
          relatedFrom: {
            include: {
              target: {
                include: {
                  scope: true
                }
              }
            }
          },
          relatedTo: {
            include: {
              source: {
                include: {
                  scope: true
                }
              }
            }
          }
        }
      });

      return issue as unknown as IssueWithRelations;
    } catch (error) {
      logger.error(`Error in findIssueById: ${error}`);
      throw error;
    }
  }

  /**
   * Create a new issue
   */
  async createIssue(data: IssueCreateInput): Promise<IssueWithRelations> {
    try {
      const {
        title,
        description,
        severity,
        issueType,
        state = 'ACTIVE',
        namespace,
        scope,
        links = []
      } = data;

      const newIssue = await prisma.issue.create({
        data: {
          title,
          description,
          severity,
          issueType,
          state,
          detectedAt: new Date(),
          namespace,
          scope: {
            create: {
              resourceType: scope.resourceType,
              resourceName: scope.resourceName,
              resourceNamespace: namespace
            }
          },
          links: {
            create: links.map(link => ({
              title: link.title,
              url: link.url
            }))
          }
        },
        include: {
          scope: true,
          links: true
        }
      });

      logger.info(`Created new issue: ${newIssue.id}`);
      return newIssue as IssueWithRelations;
    } catch (error) {
      logger.error(`Error in createIssue: ${error}`);
      throw error;
    }
  }

  /**
   * Update an existing issue
   */
  async updateIssue(id: string, data: IssueUpdateInput): Promise<IssueWithRelations> {
    try {
      const {
        title,
        description,
        severity,
        issueType,
        state,
        resolvedAt,
        links
      } = data;

      // Find the issue to verify it exists
      const existingIssue = await prisma.issue.findUnique({
        where: { id }
      });

      if (!existingIssue) {
        throw new Error(`Issue with ID ${id} not found`);
      }

      // Prepare update data
      const updateData: any = {};
      if (title !== undefined) updateData.title = title;
      if (description !== undefined) updateData.description = description;
      if (severity !== undefined) updateData.severity = severity;
      if (issueType !== undefined) updateData.issueType = issueType;
      if (state !== undefined) updateData.state = state;

      // Handle state change to RESOLVED
      if (state === 'RESOLVED' && existingIssue.state !== 'RESOLVED') {
        updateData.resolvedAt = new Date();
      } else if (resolvedAt !== undefined) {
        updateData.resolvedAt = resolvedAt;
      }

      // Start a transaction to update the issue and its links if needed
      const result = await prisma.$transaction(async (tx) => {
        // Update the issue
        const updatedIssue = await tx.issue.update({
          where: { id },
          data: updateData,
          include: {
            scope: true,
            links: true
          }
        });

        // Handle links updates if provided
        if (links !== undefined) {
          // Delete existing links
          await tx.link.deleteMany({
            where: { issueId: id }
          });

          // Create new links
          if (links.length > 0) {
            await tx.link.createMany({
              data: links.map(link => ({
                title: link.title,
                url: link.url,
                issueId: id
              }))
            });
          }

          // Get updated links
          const updatedLinks = await tx.link.findMany({
            where: { issueId: id }
          });

          updatedIssue.links = updatedLinks;
        }

        return updatedIssue;
      });

      logger.info(`Updated issue: ${id}`);
      return result as IssueWithRelations;
    } catch (error) {
      logger.error(`Error in updateIssue: ${error}`);
      throw error;
    }
  }

  /**
   * Delete an issue
   */
  async deleteIssue(id: string): Promise<void> {
    try {
      // Find the issue to get the scope ID
      const issue = await prisma.issue.findUnique({
        where: { id },
        select: { scopeId: true }
      });

      if (!issue) {
        throw new Error(`Issue with ID ${id} not found`);
      }

      // Delete in a transaction to maintain integrity
      await prisma.$transaction([
        // Delete related issues links
        prisma.relatedIssue.deleteMany({
          where: {
            OR: [
              { sourceId: id },
              { targetId: id }
            ]
          }
        }),
        // Delete links
        prisma.link.deleteMany({
          where: { issueId: id }
        }),
        // Delete the issue
        prisma.issue.delete({
          where: { id }
        }),
        // Delete the scope
        prisma.issueScope.delete({
          where: { id: issue.scopeId }
        })
      ]);

      logger.info(`Deleted issue: ${id}`);
    } catch (error) {
      logger.error(`Error in deleteIssue: ${error}`);
      throw error;
    }
  }

  /**
   * Connect two issues as related
   */
  async addRelatedIssue(sourceId: string, targetId: string): Promise<void> {
    try {
      // Check if both issues exist
      const [sourceExists, targetExists] = await Promise.all([
        prisma.issue.findUnique({ where: { id: sourceId } }),
        prisma.issue.findUnique({ where: { id: targetId } })
      ]);

      if (!sourceExists || !targetExists) {
        throw new Error('One or both issues not found');
      }

      // Check if relationship already exists
      const existingRelation = await prisma.relatedIssue.findFirst({
        where: {
          OR: [
            // Check source to target
            { sourceId, targetId },
            // Check target to source
            { sourceId: targetId, targetId: sourceId }
          ]
        }
      });

      if (existingRelation) {
        throw new Error('Relationship already exists');
      }

      // Create relationship
      await prisma.relatedIssue.create({
        data: {
          sourceId,
          targetId
        }
      });

      logger.info(`Added related issue: ${sourceId} -> ${targetId}`);
    } catch (error) {
      logger.error(`Error in addRelatedIssue: ${error}`);
      throw error;
    }
  }

  /**
   * Remove relationship between issues
   */
  async removeRelatedIssue(sourceId: string, targetId: string): Promise<void> {
    try {
      // Find the relationship
      const relation = await prisma.relatedIssue.findFirst({
        where: {
          OR: [
            // Check source to target
            { sourceId, targetId },
            // Check target to source
            { sourceId: targetId, targetId: sourceId }
          ]
        }
      });

      if (!relation) {
        throw new Error('Relationship not found');
      }

      // Delete the relationship
      await prisma.relatedIssue.delete({
        where: { id: relation.id }
      });

      logger.info(`Removed related issue: ${sourceId} <-> ${targetId}`);
    } catch (error) {
      logger.error(`Error in removeRelatedIssue: ${error}`);
      throw error;
    }
  }
}

// Export a singleton instance
const issueService = new IssueService();
export default issueService;

File: internal/services/issue_service.go
package services

import (
	"errors"
	"fmt"
	"time"

	"github.com/CryptoRodeo/kite/internal/models"
	"github.com/sirupsen/logrus"
	"gorm.io/gorm"
)

type IssueService struct {
	db     *gorm.DB       // Database instance
	logger *logrus.Logger // Logging instance
}

type IssueQueryFilters struct {
	Namespace    string
	Severity     *models.Severity
	IssueType    *models.IssueType
	State        *models.IssueState
	ResourceType string
	ResourceName string
	Search       string
	Limit        int
	Offset       int
}

type DuplicateCheckResult struct {
	IsDuplicate   bool
	ExistingIssue *models.Issue
}

func NewIssueService(db *gorm.DB, logger *logrus.Logger) *IssueService {
	return &IssueService{
		db:     db,
		logger: logger,
	}
}

// CheckForDuplicateIssue checks if a similar issue already exists
func (s *IssueService) CheckForDuplicateIssue(req models.CreateIssueRequest) (*DuplicateCheckResult, error) {
	var existingIssue models.Issue
	err := s.db.
		Preload("Links").
		Joins("JOIN issue_scopes on issues.scope_id = issue_scopes.id").
		Where("issues.namespace = ? AND issues.issue_type = ? AND issues.state = ?",
			req.Namespace, req.IssueType, models.IssueStateActive).
		Where("issue_scopes.resource_type = ? AND issue_scopes.resource_name = ? AND issue_scopes.resource_namespace = ?",
			req.Scope.ResourceType, req.Scope.ResourceName, req.Namespace).
		First(&existingIssue).Error
	if err != nil {
		// Check if the error is no record was found.
		// If it is, the issue is not a duplicate.
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return &DuplicateCheckResult{IsDuplicate: false}, nil
		}
		s.logger.WithError(err).Error("Failed to check for duplicate issues")
		return nil, fmt.Errorf("faield to check for duplicates: %w", err)
	}
	s.logger.WithField("existing_issue_id", existingIssue.ID).Info("found duplicate")
	return &DuplicateCheckResult{
		IsDuplicate:   true,
		ExistingIssue: &existingIssue,
	}, nil
}

// FindIssues retrieves issues with optional filters
func (s *IssueService) FindIssues(filters IssueQueryFilters) (*models.IssueResponse, error) {
	var issues []models.Issue
	var total int64

	// Build base query
	// Preload any associations
	query := s.db.Model(&models.Issue{}).
		Preload("Scope").
		Preload("Links").
		Preload("RelatedFrom.Target.Scope").
		Preload("RelatedTo.Source.Scope")

	// Apply filters to the database query
	if filters.Namespace != "" {
		query = query.Where("namespace = ?", filters.Namespace)
	}
	if filters.Severity != nil {
		query = query.Where("severity = ?", *filters.Severity)
	}
	if filters.IssueType != nil {
		query = query.Where("issue_type = ?", *filters.IssueType)
	}
	if filters.State != nil {
		query = query.Where("state = ?", *filters.State)
	}
	if filters.ResourceType != "" {
		query = query.Joins("JOIN issue_scopes ON issues.scope_id = issue_scopes.id").
			Where("issue_scopes.resource_type = ?", filters.ResourceType)
	}
	if filters.ResourceName != "" {
		query = query.Joins("JOIN issue_scopes ON issues.scope_id = issue_scopes.id").
			Where("issue_scopes.resource_name = ?", filters.ResourceName)
	}
	if filters.Search != "" {
		searchPattern := "%" + filters.Search + "%"
		query = query.Where("title ILIKE ? OR description ILIKE ?", searchPattern, searchPattern)
	}

	// Get total count for pagination
	if err := query.Count(&total).Error; err != nil {
		s.logger.WithError(err).Error("Failed to count issues")
		return nil, fmt.Errorf("failed to count issues: %w", err)
	}

	// Apply pagination and ordering
	if filters.Limit == 0 {
		filters.Limit = 50
	}

	if err := query.Order("detected_at DESC").
		Offset(filters.Offset).
		Limit(filters.Limit).
		Find(&issues).
		Error; err != nil {
		s.logger.WithError(err).Error("Failed to find issues")
		return nil, fmt.Errorf("failed to find issues: %w", err)
	}

	return &models.IssueResponse{
		Data:   issues,
		Total:  total,
		Limit:  filters.Limit,
		Offset: filters.Offset,
	}, nil
}

// FindIssueByID retrieves a single issue by ID
func (s *IssueService) FindIssueByID(id string) (*models.Issue, error) {
	var issue models.Issue

	// Find issue, load associations
	err := s.db.
		Preload("Scope").
		Preload("RelatedFrom.Target.Scope").
		Preload("RelatedTo.Source.Scope").
		First(&issue, "id = ?", id).Error

	if err != nil {
		// Check if the error is record not found
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, nil
		}
		s.logger.WithError(err).WithField("issue_id", id).Error("Failed to find issue by ID")
		return nil, fmt.Errorf("failed to find issue: %w", err)
	}

	return &issue, nil
}

// CreateIssue creates a new issue
func (s *IssueService) CreateIssue(req models.CreateIssueRequest) (*models.Issue, error) {
	// check for duplicates
	duplicateResult, err := s.CheckForDuplicateIssue(req)
	if err != nil {
		return nil, fmt.Errorf("failed to check for duplicates: %w", err)
	}

	// Check if this issue is a duplicate.
	if duplicateResult.IsDuplicate && duplicateResult.ExistingIssue != nil {
		// Update existing issue instead of creating a new one
		updateReq := models.UpdateIssueRequest{
			Title:       &req.Title,
			Description: &req.Description,
			Severity:    &req.Severity,
			IssueType:   &req.IssueType,
		}
		if req.State != "" {
			updateReq.State = &req.State
		}
		return s.UpdateIssue(duplicateResult.ExistingIssue.ID, updateReq)
	}

	// Create new issue
	now := time.Now()
	state := req.State
	// Assume the state of the issue is active if not sent in request
	if state == "" {
		state = models.IssueStateActive
	}

	// Set resource namespace to match issue namespace if not provided
	resourceNamespace := req.Scope.ResourceName
	if resourceNamespace == "" {
		resourceNamespace = req.Namespace
	}

	issue := models.Issue{
		Title:       req.Title,
		Description: req.Description,
		Severity:    req.Severity,
		IssueType:   req.IssueType,
		State:       req.State,
		DetectedAt:  now,
		Namespace:   req.Namespace,
		Scope: models.IssueScope{
			ResourceType:      req.Scope.ResourceType,
			ResourceName:      req.Scope.ResourceName,
			ResourceNamespace: req.Scope.ResourceNamespace,
		},
	}

	// Convert any links
	for _, linkReq := range req.Links {
		issue.Links = append(issue.Links, models.Link{
			Title: linkReq.Title,
			URL:   linkReq.URL,
		})
	}

	// Create in a transaction
	err = s.db.Transaction(func(tx *gorm.DB) error {
		if err := tx.Create(&issue).Error; err != nil {
			return fmt.Errorf("failed to create issue: %w", err)
		}
		return nil
	})

	if err != nil {
		s.logger.WithError(err).Error("Failed to create issue")
		return nil, err
	}

	s.logger.WithField("issue_id", issue.ID).Info("Created new issue")

	// Reload with associations
	return s.FindIssueByID(issue.ID)
}

// UpdateIssue updates and existing issue
func (s *IssueService) UpdateIssue(id string, req models.UpdateIssueRequest) (*models.Issue, error) {
	// Find existing issue
	existingIssue, err := s.FindIssueByID(id)
	if err != nil {
		return nil, err
	}
	if existingIssue == nil {
		return nil, fmt.Errorf("issue with ID %s not found", id)
	}

	// Prepare updates
	updates := map[string]interface{}{
		"updated_at": time.Now(),
	}

	if req.Title != nil {
		updates["title"] = *req.Title
	}
	if req.Description != nil {
		updates["description"] = *req.Description
	}
	if req.Severity != nil {
		updates["severity"] = *req.Severity
	}
	if req.IssueType != nil {
		updates["issue_type"] = *req.IssueType
	}
	if req.State != nil {
		updates["state"] = *req.State
		// Handle state change to RESOLVED
		if *req.State == models.IssueStateResolved && existingIssue.State != models.IssueStateResolved {
			now := time.Now()
			// Add time when issue was resolved
			updates["resolved_at"] = &now
		}
	}
	if req.ResolvedAt != nil {
		updates["resolved_at"] = req.ResolvedAt
	}

	// Perform updates in a transaction
	// Update issue first, then links (if any)
	err = s.db.Transaction(func(tx *gorm.DB) error {
		// Update issue
		if err := tx.Model(&existingIssue).Updates(updates).Error; err != nil {
			return fmt.Errorf("failed to update issue: %w", err)
		}

		// Handle link updates if provided
		if req.Links != nil {
			// Delete old links
			if err := tx.Where("issue_id = ?", id).Delete(&models.Link{}).Error; err != nil {
				return fmt.Errorf("failed to delete old links: %w", err)
			}

			// Create new links
			for _, linkReq := range req.Links {
				link := models.Link{
					Title:   linkReq.Title,
					URL:     linkReq.URL,
					IssueID: id,
				}
				if err := tx.Create(&link).Error; err != nil {
					return fmt.Errorf("failed to create link: %w", err)
				}
			}
		}
		return nil
	})

	if err != nil {
		s.logger.WithError(err).WithField("issue_id", id).Error("Failed to update issue")
		return nil, err
	}

	s.logger.WithField("issue_id", id).Info("Updated issue")

	return s.FindIssueByID(id)
}

// DeleteIssue deletes an issue and related entities
func (s *IssueService) DeleteIssue(id string) error {
	// Find the issue to get scope ID
	issue, err := s.FindIssueByID(id)
	if err != nil {
		return err
	}
	if issue == nil {
		return fmt.Errorf("issue with ID %s not found", id)
	}

	// Delete in transaction so we have control of the order
	err = s.db.Transaction(func(tx *gorm.DB) error {
		// Delete related issue relationships first using issue id
		if err := tx.Where("source_id = ? OR target_id = ?", id, id).Delete(&models.RelatedIssue{}).Error; err != nil {
			return fmt.Errorf("failed to delete related issues: %w", err)
		}

		// Delete links by issue id
		if err := tx.Where("issue_id = ?", id).Delete(&models.Link{}).Error; err != nil {
			return fmt.Errorf("failed to delete links: %w", err)
		}

		// Delete the issue by id
		if err := tx.Delete(&models.Issue{}, "id = ?", id).Error; err != nil {
			return fmt.Errorf("failed to delete issue: %w", err)
		}

		// Delete the issue scope by scope id
		if err := tx.Delete(&models.IssueScope{}, "id = ?", issue.ScopeID).Error; err != nil {
			return fmt.Errorf("failed to delete issue scope: %w", err)
		}

		return nil
	})

	if err != nil {
		s.logger.WithError(err).WithField("issue_id", id).Error("failed to delete issue")
		return err
	}

	s.logger.WithField("issue_id", id).Info("Deleted issue")
	return nil
}

// AddRelatedIsue creates a relationship between two issues
func (s *IssueService) AddRelatedIssue(sourceID, targetID string) error {
	// Check if both issues exist
	source, err := s.FindIssueByID(sourceID)
	if err != nil {
		return err
	}
	target, err := s.FindIssueByID(targetID)
	if err != nil {
		return err
	}
	if source == nil || target == nil {
		return errors.New("one or both issues not found")
	}

	// Check if relationship already exists
	var existingRelation models.RelatedIssue
	err = s.db.Where("(source_id = ? AND target_id = ?) OR (source_id = ? AND target_id = ?)",
		sourceID, targetID, targetID, sourceID).First(&existingRelation).Error

	if err == nil {
		return errors.New("relationship already exists")
	}
	// Check if we get any other error besides Record Not Found
	if !errors.Is(err, gorm.ErrRecordNotFound) {
		return fmt.Errorf("failed to check exiting relationship: %w", err)
	}

	// Create relationship
	relation := models.RelatedIssue{
		SourceID: sourceID,
		TargetID: targetID,
	}

	if err := s.db.Create(&relation).Error; err != nil {
		s.logger.WithError(err).Error("Failed to add related issue")
		return fmt.Errorf("failed to create relationship: %w", err)
	}

	s.logger.WithFields(logrus.Fields{
		"source_id": sourceID,
		"target_id": targetID,
	}).Info("Added related issue")
	return nil
}

// RemoveRelatedIssue removes a relationship between issues
func (s *IssueService) RemoveRelatedIssue(sourceID, targetID string) error {
	result := s.db.Where("(source_id = ? AND target_id = ?) OR (source_id = ? AND target_id = ?)",
		sourceID, targetID, targetID, sourceID).Delete(&models.RelatedIssue{})

	if result.Error != nil {
		s.logger.WithError(result.Error).Error("failed to remove related issue")
		return fmt.Errorf("failed to remove relationship: %w", result.Error)
	}

	if result.RowsAffected == 0 {
		return errors.New("relationship not found")
	}

	s.logger.WithFields(logrus.Fields{
		"source_id": sourceID,
		"target_id": targetID,
	}).Info("Removed related issue")

	return nil
}

// ResolveIssuesByScope resolves all active issues for a given scope
func (s *IssueService) ResolveIssuesByScope(resourceType, resourceName, namespace string) (int64, error) {
	now := time.Now()

	result := s.db.Model(&models.Issue{}).
		Joins("JOIN issue_scopes ON issues.scope_id = issue_scopes.id").
		Where("issues.state = ? AND issues.namespace = ?", models.IssueStateActive, namespace).
		Where("issue_scopes.resource_type = ? AND issue_scopes.resource_name = ?", resourceType, resourceName).
		Updates(map[string]any{
			"state":       models.IssueStateResolved,
			"resolved_at": &now,
			"updated_at":  now,
		})

	if result.Error != nil {
		s.logger.WithError(result.Error).Error("Failed to resolve issues by scope")
		return 0, fmt.Errorf("failed to resolve issues: %w", result.Error)
	}

	count := result.RowsAffected
	s.logger.WithFields(logrus.Fields{
		"resource_type": resourceType,
		"resource_name": resourceName,
		"namespace":     namespace,
		"count":         count,
	}).Info("Resolved issues by scope")

	return count, nil
}
